File: remix-of-vesper-blog-main/src/services/shipmentService.ts
// Shipment service — 3-way logistics workflow
// Status: awaiting_logistics → logistics_proposed → approved → in_transit → delivered
import {
    collection,
    doc,
    addDoc,
    getDoc,
    getDocs,
    updateDoc,
    query,
    where,
    serverTimestamp,
} from "firebase/firestore";
import { db, auth } from "@/firebase";

// -------- Types --------

export type ShipmentStatus =
    | "awaiting_logistics"
    | "logistics_proposed"
    | "approved"
    | "in_transit"
    | "delivered";

export interface ShipmentDocument {
    id: string;
    commitmentId: string;
    cropId: string;
    cropName: string;
    itemType: "primary" | "byproduct";
    itemName: string;
    farmerId: string;
    buyerId: string;
    pickupLocation: string;
    deliveryLocation: string;
    quantity: number;
    logisticsProviderId: string | null;
    status: ShipmentStatus;
    // Route planning fields
    proposedRoute: string;
    estimatedCost: number | null;
    distanceKm: number | null;
    estimatedDuration: string;
    pickupCoordinates?: { lat: number; lng: number } | null;
    deliveryCoordinates?: { lat: number; lng: number } | null;
    routePolyline?: { lat: number; lng: number }[] | null;
    createdAt: Date | null;
    updatedAt: Date | null;
}

// Valid status transitions (logistics-controlled)
const VALID_TRANSITIONS: Record<string, string[]> = {
    awaiting_logistics: ["logistics_proposed"],
    logistics_proposed: ["approved", "awaiting_logistics"], // buyer approve/reject
    approved: ["in_transit"],
    in_transit: ["delivered"],
    delivered: [],
};

// -------- Helpers --------

function getUid(): string | null {
    let uid = auth.currentUser?.uid;
    if (!uid) {
        try {
            const s = sessionStorage.getItem("vasudha_user");
            if (s) uid = JSON.parse(s).id;
        } catch { /* ignore */ }
    }
    return uid || null;
}

function mapDoc(d: any): ShipmentDocument {
    const data = d.data ? d.data() : d;
    return {
        id: d.id || "",
        commitmentId: data.commitmentId || "",
        cropId: data.cropId || "",
        cropName: data.cropName || "",
        itemType: data.itemType || "primary",
        itemName: data.itemName || data.cropName || "",
        farmerId: data.farmerId || "",
        buyerId: data.buyerId || "",
        pickupLocation: data.pickupLocation || "",
        deliveryLocation: data.deliveryLocation || "",
        quantity: data.quantity || 0,
        logisticsProviderId: data.logisticsProviderId || null,
        status: (data.status || "awaiting_logistics") as ShipmentStatus,
        proposedRoute: data.proposedRoute || "",
        estimatedCost: data.estimatedCost ?? null,
        distanceKm: data.distanceKm ?? null,
        estimatedDuration: data.estimatedDuration || "",
        pickupCoordinates: data.pickupCoordinates ?? null,
        deliveryCoordinates: data.deliveryCoordinates ?? null,
        routePolyline: data.routePolyline ?? null,
        createdAt: data.createdAt?.toDate?.() || null,
        updatedAt: data.updatedAt?.toDate?.() || null,
    };
}

// -------- Create Shipment (auto on commitment accept) --------

export async function createShipmentFromCommitment(commitment: {
    id: string;
    cropId: string;
    cropName: string;
    itemType?: "primary" | "byproduct";
    itemName?: string;
    farmerId: string;
    buyerId: string;
    quantity: number;
}): Promise<string> {
    try {
        // Guard: prevent duplicate
        const existingQ = query(
            collection(db, "shipments"),
            where("commitmentId", "==", commitment.id)
        );
        const existingSnap = await getDocs(existingQ);
        if (!existingSnap.empty) {
            console.log("[ShipmentService] Shipment already exists for commitment:", commitment.id);
            return existingSnap.docs[0].id;
        }

        // Read crop for pickup location
        let pickupLocation = "";
        try {
            const cropSnap = await getDoc(doc(db, "crops", commitment.cropId));
            if (cropSnap.exists()) {
                pickupLocation = cropSnap.data().location || "";
            }
        } catch (err) {
            console.warn("[ShipmentService] Could not read crop location:", err);
        }

        // Try to find delivery location from buyer's demand
        let deliveryLocation = "";
        try {
            const demandQ = query(
                collection(db, "demands"),
                where("buyerId", "==", commitment.buyerId),
                where("cropName", "==", commitment.cropName.toLowerCase())
            );
            const demandSnap = await getDocs(demandQ);
            if (!demandSnap.empty) {
                deliveryLocation = demandSnap.docs[0].data().location || "";
            }
        } catch (err) {
            console.warn("[ShipmentService] Could not read demand location:", err);
        }

        const shipmentData = {
            commitmentId: commitment.id,
            cropId: commitment.cropId,
            cropName: commitment.cropName,
            itemType: commitment.itemType || "primary",
            itemName: commitment.itemName || commitment.cropName,
            farmerId: commitment.farmerId,
            buyerId: commitment.buyerId,
            pickupLocation,
            deliveryLocation,
            quantity: commitment.quantity,
            logisticsProviderId: null,
            status: "awaiting_logistics" as ShipmentStatus,
            proposedRoute: "",
            estimatedCost: null,
            distanceKm: null,
            estimatedDuration: "",
            createdAt: serverTimestamp(),
            updatedAt: serverTimestamp(),
        };

        const docRef = await addDoc(collection(db, "shipments"), shipmentData);
        console.log("[ShipmentService] Created shipment:", docRef.id);
        return docRef.id;
    } catch (err) {
        console.error("[ShipmentService] createShipmentFromCommitment failed:", err);
        throw err;
    }
}

// -------- Logistics Propose (awaiting_logistics → logistics_proposed) --------

export async function proposeShipment(
    shipmentId: string,
    proposedRoute: string,
    estimatedCost: number
): Promise<void> {
    const uid = getUid();
    if (!uid) throw new Error("Not authenticated");

    try {
        const shipRef = doc(db, "shipments", shipmentId);
        const shipSnap = await getDoc(shipRef);

        if (!shipSnap.exists()) throw new Error("Shipment not found");

        const shipment = shipSnap.data();
        if (shipment.status !== "awaiting_logistics") {
            throw new Error("Shipment is not awaiting logistics");
        }
        if (shipment.logisticsProviderId && shipment.logisticsProviderId !== uid) {
            throw new Error("Another logistics provider already proposed");
        }

        // Mock route data (would use Google Maps API if key available)
        const distanceKm = Math.round(Math.random() * 300 + 50);
        const hours = Math.round(distanceKm / 45);
        const estimatedDuration = `${hours}h ${Math.round(Math.random() * 59)}m`;

        await updateDoc(shipRef, {
            logisticsProviderId: uid,
            status: "logistics_proposed",
            proposedRoute,
            estimatedCost,
            distanceKm,
            estimatedDuration,
            updatedAt: serverTimestamp(),
        });

        console.log("[ShipmentService] Proposed shipment", shipmentId);
    } catch (err) {
        console.error("[ShipmentService] proposeShipment failed:", err);
        throw err;
    }
}

// -------- Buyer Approve/Reject (logistics_proposed → approved | awaiting_logistics) --------

export async function approveShipment(
    shipmentId: string,
    approved: boolean
): Promise<void> {
    const uid = getUid();
    if (!uid) throw new Error("Not authenticated");

    try {
        const shipRef = doc(db, "shipments", shipmentId);
        const shipSnap = await getDoc(shipRef);

        if (!shipSnap.exists()) throw new Error("Shipment not found");

        const shipment = shipSnap.data();
        if (shipment.status !== "logistics_proposed") {
            throw new Error("Shipment is not awaiting buyer approval");
        }
        if (uid !== shipment.buyerId) {
            throw new Error("Only the buyer can approve/reject shipments");
        }

        if (approved) {
            await updateDoc(shipRef, {
                status: "approved",
                updatedAt: serverTimestamp(),
            });
            console.log("[ShipmentService] Buyer approved shipment", shipmentId);
        } else {
            // Reject → reset to awaiting_logistics
            await updateDoc(shipRef, {
                status: "awaiting_logistics",
                logisticsProviderId: null,
                proposedRoute: "",
                estimatedCost: null,
                distanceKm: null,
                estimatedDuration: "",
                updatedAt: serverTimestamp(),
            });
            console.log("[ShipmentService] Buyer rejected shipment", shipmentId);
        }
    } catch (err) {
        console.error("[ShipmentService] approveShipment failed:", err);
        throw err;
    }
}

// -------- Update Status (approved → in_transit → delivered) --------

export async function updateShipmentStatus(
    shipmentId: string,
    newStatus: ShipmentStatus
): Promise<void> {
    const uid = getUid();
    if (!uid) throw new Error("Not authenticated");

    try {
        const shipRef = doc(db, "shipments", shipmentId);
        const shipSnap = await getDoc(shipRef);

        if (!shipSnap.exists()) throw new Error("Shipment not found");

        const shipment = shipSnap.data();
        const currentStatus = shipment.status as ShipmentStatus;

        const allowed = VALID_TRANSITIONS[currentStatus] || [];
        if (!allowed.includes(newStatus)) {
            throw new Error(`Cannot transition from "${currentStatus}" to "${newStatus}"`);
        }

        await updateDoc(shipRef, {
            status: newStatus,
            updatedAt: serverTimestamp(),
        });

        // delivered → mark commitment completed
        if (newStatus === "delivered" && shipment.commitmentId) {
            try {
                const commitRef = doc(db, "commitments", shipment.commitmentId);
                const commitSnap = await getDoc(commitRef);
                if (commitSnap.exists() && commitSnap.data().status === "accepted") {
                    await updateDoc(commitRef, {
                        status: "completed",
                        updatedAt: serverTimestamp(),
                    });
                    console.log("[ShipmentService] Commitment", shipment.commitmentId, "→ completed");
                }
            } catch (err) {
                console.error("[ShipmentService] Failed to update commitment:", err);
            }
        }

        console.log("[ShipmentService] Updated shipment", shipmentId, "→", newStatus);
    } catch (err) {
        console.error("[ShipmentService] updateShipmentStatus failed:", err);
        throw err;
    }
}

// -------- Query Functions --------

/** Shipments waiting for logistics (for logistics dashboard) */
export async function getAvailableShipments(): Promise<ShipmentDocument[]> {
    try {
        const q = query(collection(db, "shipments"), where("status", "==", "awaiting_logistics"));
        const snap = await getDocs(q);
        return snap.docs.map(mapDoc);
    } catch (err) {
        console.error("[ShipmentService] getAvailableShipments failed:", err);
        return [];
    }
}

/** Shipments proposed by current logistics provider */
export async function getMyShipments(): Promise<ShipmentDocument[]> {
    const uid = getUid();
    if (!uid) return [];
    try {
        const q = query(collection(db, "shipments"), where("logisticsProviderId", "==", uid));
        const snap = await getDocs(q);
        return snap.docs.map(mapDoc);
    } catch (err) {
        console.error("[ShipmentService] getMyShipments failed:", err);
        return [];
    }
}

/** Shipments awaiting buyer approval (logistics_proposed for this buyer) */
export async function getShipmentsAwaitingApproval(): Promise<ShipmentDocument[]> {
    const uid = getUid();
    if (!uid) return [];
    try {
        const q = query(
            collection(db, "shipments"),
            where("buyerId", "==", uid),
            where("status", "==", "logistics_proposed")
        );
        const snap = await getDocs(q);
        return snap.docs.map(mapDoc);
    } catch (err) {
        console.error("[ShipmentService] getShipmentsAwaitingApproval failed:", err);
        return [];
    }
}

/** Shipments for a farmer */
export async function getShipmentsByFarmer(): Promise<ShipmentDocument[]> {
    const uid = getUid();
    if (!uid) return [];
    try {
        const q = query(collection(db, "shipments"), where("farmerId", "==", uid));
        const snap = await getDocs(q);
        return snap.docs.map(mapDoc);
    } catch (err) {
        console.error("[ShipmentService] getShipmentsByFarmer failed:", err);
        return [];
    }
}

/** Shipments for a buyer */
export async function getShipmentsByBuyer(): Promise<ShipmentDocument[]> {
    const uid = getUid();
    if (!uid) return [];
    try {
        const q = query(collection(db, "shipments"), where("buyerId", "==", uid));
        const snap = await getDocs(q);
        return snap.docs.map(mapDoc);
    } catch (err) {
        console.error("[ShipmentService] getShipmentsByBuyer failed:", err);
        return [];
    }
}

/** Get a single shipment by ID */
export async function getShipmentById(shipmentId: string): Promise<ShipmentDocument | null> {
    try {
        const shipSnap = await getDoc(doc(db, "shipments", shipmentId));
        if (!shipSnap.exists()) return null;
        return mapDoc(shipSnap);
    } catch (err) {
        console.error("[ShipmentService] getShipmentById failed:", err);
        return null;
    }
}

// -------- Indian city coordinate lookup for route generation --------

const CITY_COORDS: Record<string, { lat: number; lng: number }> = {
    "mumbai": { lat: 19.0760, lng: 72.8777 },
    "delhi": { lat: 28.7041, lng: 77.1025 },
    "new delhi": { lat: 28.6139, lng: 77.2090 },
    "bangalore": { lat: 12.9716, lng: 77.5946 },
    "bengaluru": { lat: 12.9716, lng: 77.5946 },
    "hyderabad": { lat: 17.3850, lng: 78.4867 },
    "chennai": { lat: 13.0827, lng: 80.2707 },
    "kolkata": { lat: 22.5726, lng: 88.3639 },
    "pune": { lat: 18.5204, lng: 73.8567 },
    "ahmedabad": { lat: 23.0225, lng: 72.5714 },
    "jaipur": { lat: 26.9124, lng: 75.7873 },
    "surat": { lat: 21.1702, lng: 72.8311 },
    "lucknow": { lat: 26.8467, lng: 80.9462 },
    "kanpur": { lat: 26.4499, lng: 80.3319 },
    "nagpur": { lat: 21.1458, lng: 79.0882 },
    "indore": { lat: 22.7196, lng: 75.8577 },
    "thane": { lat: 19.2183, lng: 72.9781 },
    "bhopal": { lat: 23.2599, lng: 77.4126 },
    "visakhapatnam": { lat: 17.6868, lng: 83.2185 },
    "vizag": { lat: 17.6868, lng: 83.2185 },
    "pimpri-chinchwad": { lat: 18.6279, lng: 73.8009 },
    "patna": { lat: 25.6093, lng: 85.1376 },
    "vadodara": { lat: 22.3072, lng: 73.1812 },
    "ghaziabad": { lat: 28.6692, lng: 77.4538 },
    "ludhiana": { lat: 30.9010, lng: 75.8573 },
    "agra": { lat: 27.1767, lng: 78.0081 },
    "nashik": { lat: 19.9975, lng: 73.7898 },
    "faridabad": { lat: 28.4089, lng: 77.3178 },
    "meerut": { lat: 28.9845, lng: 77.7064 },
    "rajkot": { lat: 22.3039, lng: 70.8022 },
    "varanasi": { lat: 25.3176, lng: 82.9739 },
    "coimbatore": { lat: 11.0168, lng: 76.9558 },
    "srinagar": { lat: 34.0837, lng: 74.7973 },
    "madurai": { lat: 9.9252, lng: 78.1198 },
    "vijayawada": { lat: 16.5062, lng: 80.6480 },
    "chandigarh": { lat: 30.7333, lng: 76.7794 },
    "mysore": { lat: 12.2958, lng: 76.6394 },
    "mysuru": { lat: 12.2958, lng: 76.6394 },
    "ranchi": { lat: 23.3441, lng: 85.3096 },
    "raipur": { lat: 21.2514, lng: 81.6296 },
    "guwahati": { lat: 26.1445, lng: 91.7362 },
    "kochi": { lat: 9.9312, lng: 76.2673 },
    "cochin": { lat: 9.9312, lng: 76.2673 },
    "thiruvananthapuram": { lat: 8.5241, lng: 76.9366 },
    "trivandrum": { lat: 8.5241, lng: 76.9366 },
    "dehradun": { lat: 30.3165, lng: 78.0322 },
    "mangalore": { lat: 12.9141, lng: 74.8560 },
    "amritsar": { lat: 31.6340, lng: 74.8723 },
    "noida": { lat: 28.5355, lng: 77.3910 },
    "gurgaon": { lat: 28.4595, lng: 77.0266 },
    "gurugram": { lat: 28.4595, lng: 77.0266 },
    "jodhpur": { lat: 26.2389, lng: 73.0243 },
    "udaipur": { lat: 24.5854, lng: 73.7125 },
    "warangal": { lat: 17.9784, lng: 79.5941 },
    "guntur": { lat: 16.3067, lng: 80.4365 },
    "tirupati": { lat: 13.6288, lng: 79.4192 },
    "nellore": { lat: 14.4426, lng: 79.9865 },
    "kakinada": { lat: 16.9891, lng: 82.2475 },
    "hubli": { lat: 15.3647, lng: 75.1240 },
    "telangana": { lat: 17.3850, lng: 78.4867 },
};

/** Find coordinates for a location string (fuzzy match against city names) */
function geocodeCity(location: string): { lat: number; lng: number } | null {
    if (!location) return null;
    const lower = location.toLowerCase().trim();
    // Exact match
    if (CITY_COORDS[lower]) return CITY_COORDS[lower];
    // Partial match — check if any city name is contained in the location string
    for (const [city, coords] of Object.entries(CITY_COORDS)) {
        if (lower.includes(city) || city.includes(lower)) return coords;
    }
    return null;
}

/** Haversine distance between two lat/lng points in km */
function haversineKm(a: { lat: number; lng: number }, b: { lat: number; lng: number }): number {
    const R = 6371;
    const dLat = (b.lat - a.lat) * Math.PI / 180;
    const dLng = (b.lng - a.lng) * Math.PI / 180;
    const sinLat = Math.sin(dLat / 2);
    const sinLng = Math.sin(dLng / 2);
    const h = sinLat * sinLat + Math.cos(a.lat * Math.PI / 180) * Math.cos(b.lat * Math.PI / 180) * sinLng * sinLng;
    return Math.round(R * 2 * Math.atan2(Math.sqrt(h), Math.sqrt(1 - h)));
}

/** Generate route data for a shipment (geocode cities → real coordinates + polyline) */
export async function generateRoute(shipmentId: string, force = false): Promise<ShipmentDocument> {
    try {
        const shipRef = doc(db, "shipments", shipmentId);
        const shipSnap = await getDoc(shipRef);

        if (!shipSnap.exists()) throw new Error("Shipment not found");
        const shipment = shipSnap.data();

        // If force mode, clear old route data first so it can be regenerated
        if (force && (shipment.pickupCoordinates || shipment.routePolyline)) {
            await updateDoc(shipRef, {
                pickupCoordinates: null,
                deliveryCoordinates: null,
                routePolyline: null,
                updatedAt: serverTimestamp(),
            });
            console.log("[ShipmentService] Cleared old route data for", shipmentId);
        } else if (!force && shipment.pickupCoordinates && shipment.deliveryCoordinates && shipment.routePolyline) {
            // Prevent double generation (only if not forcing)
            console.log("[ShipmentService] Route already generated for", shipmentId);
            return mapDoc(shipSnap);
        }

        // Geocode pickup and delivery from location names
        const pickupCoords = geocodeCity(shipment.pickupLocation || "");
        const deliveryCoords = geocodeCity(shipment.deliveryLocation || "");

        if (!pickupCoords || !deliveryCoords) {
            throw new Error(
                `Cannot geocode locations: pickup="${shipment.pickupLocation || "?"}", delivery="${shipment.deliveryLocation || "?"}". ` +
                `Add these cities to the lookup or enter a recognized Indian city name.`
            );
        }

        // Parse "via" waypoints from proposedRoute (e.g. "via hyderabad", "via nh65", "via nagpur, pune")
        const viaWaypoints: { lat: number; lng: number }[] = [];
        if (shipment.proposedRoute) {
            const routeStr = (shipment.proposedRoute as string).toLowerCase();
            // Extract city names after "via" keyword
            const viaMatch = routeStr.replace(/^via\s+/i, "");
            // Split on commas, "and", slashes, hyphens for multiple stops
            const parts = viaMatch.split(/[,\/&]+|\band\b/).map((s: string) => s.trim()).filter(Boolean);
            for (const part of parts) {
                const coords = geocodeCity(part);
                if (coords) viaWaypoints.push(coords);
            }
        }

        // Build ordered list of key stops: Pickup → Via stops → Delivery
        const keyStops = [pickupCoords, ...viaWaypoints, deliveryCoords];

        // Compute total road distance along all segments
        let totalStraightKm = 0;
        for (let i = 0; i < keyStops.length - 1; i++) {
            totalStraightKm += haversineKm(keyStops[i], keyStops[i + 1]);
        }
        const roadDistance = Math.round(totalStraightKm * 1.3);
        const avgSpeedKmh = 45;
        const totalHours = roadDistance / avgSpeedKmh;
        const hours = Math.floor(totalHours);
        const mins = Math.round((totalHours - hours) * 60);
        const estimatedDuration = `${hours}h ${mins}m`;

        // Generate polyline through all key stops with interpolation between each segment
        const routePolyline: { lat: number; lng: number }[] = [];
        for (let seg = 0; seg < keyStops.length - 1; seg++) {
            const from = keyStops[seg];
            const to = keyStops[seg + 1];
            const pointsInSegment = 4 + Math.floor(Math.random() * 3);
            for (let i = 0; i <= pointsInSegment; i++) {
                // Skip duplicate points at segment boundaries (except first segment start)
                if (seg > 0 && i === 0) continue;
                const frac = i / pointsInSegment;
                const dLat = to.lat - from.lat;
                const dLng = to.lng - from.lng;
                const len = Math.sqrt(dLat * dLat + dLng * dLng) + 0.001;
                // Slight perpendicular offset for road-curve feel (not at endpoints)
                const offset = (i === 0 || i === pointsInSegment) ? 0 : (Math.random() - 0.5) * 0.3;
                routePolyline.push({
                    lat: +(from.lat + frac * dLat + offset * (-dLng / len)).toFixed(6),
                    lng: +(from.lng + frac * dLng + offset * (dLat / len)).toFixed(6),
                });
            }
        }

        await updateDoc(shipRef, {
            pickupCoordinates: pickupCoords,
            deliveryCoordinates: deliveryCoords,
            routePolyline,
            distanceKm: roadDistance,
            estimatedDuration,
            updatedAt: serverTimestamp(),
        });

        const viaNames = viaWaypoints.length > 0 ? ` via ${shipment.proposedRoute}` : "";
        console.log("[ShipmentService] Generated route for", shipmentId, `${shipment.pickupLocation} →${viaNames} ${shipment.deliveryLocation}`, roadDistance, "km");
        const updatedSnap = await getDoc(shipRef);
        return mapDoc(updatedSnap);
    } catch (err) {
        console.error("[ShipmentService] generateRoute failed:", err);
        throw err;
    }
}

File: backend/app/services/shipment_service.py
"""
Shipment service — 3-way logistics workflow backend.
Status: awaiting_logistics → logistics_proposed → approved → in_transit → delivered
"""
import logging
import random
from firebase_admin import firestore
from app.firebase_init import get_firestore_client

logger = logging.getLogger(__name__)

VALID_TRANSITIONS = {
    "awaiting_logistics": ["logistics_proposed"],
    "logistics_proposed": ["approved", "awaiting_logistics"],
    "approved": ["in_transit"],
    "in_transit": ["delivered"],
    "delivered": [],
}


def create_shipment_from_commitment(commitment: dict) -> str:
    """Create a shipment when a commitment is accepted."""
    try:
        db = get_firestore_client()
        existing = db.collection("shipments").where(
            "commitmentId", "==", commitment["id"]
        ).get()
        if existing:
            logger.info(f"Shipment already exists for commitment {commitment['id']}")
            return existing[0].id

        pickup_location = ""
        try:
            crop_doc = db.collection("crops").document(commitment["cropId"]).get()
            if crop_doc.exists:
                pickup_location = crop_doc.to_dict().get("location", "")
        except Exception as e:
            logger.warning(f"Could not read crop location: {e}")

        delivery_location = ""
        try:
            demands = db.collection("demands").where(
                "buyerId", "==", commitment["buyerId"]
            ).where(
                "cropName", "==", commitment.get("cropName", "").lower()
            ).get()
            if demands:
                delivery_location = demands[0].to_dict().get("location", "")
        except Exception as e:
            logger.warning(f"Could not read demand location: {e}")

        shipment_data = {
            "commitmentId": commitment["id"],
            "cropId": commitment["cropId"],
            "cropName": commitment.get("cropName", ""),
            "itemType": commitment.get("itemType", "primary"),
            "itemName": commitment.get("itemName", commitment.get("cropName", "")),
            "farmerId": commitment["farmerId"],
            "buyerId": commitment["buyerId"],
            "pickupLocation": pickup_location,
            "deliveryLocation": delivery_location,
            "quantity": commitment.get("quantity", 0),
            "logisticsProviderId": None,
            "status": "awaiting_logistics",
            "proposedRoute": "",
            "estimatedCost": None,
            "distanceKm": None,
            "estimatedDuration": "",
            "createdAt": firestore.SERVER_TIMESTAMP,
            "updatedAt": firestore.SERVER_TIMESTAMP,
        }


        _, doc_ref = db.collection("shipments").add(shipment_data)
        logger.info(f"Created shipment {doc_ref.id} for commitment {commitment['id']}")
        return doc_ref.id

    except Exception as e:
        logger.error(f"create_shipment_from_commitment failed: {e}")
        raise


def propose_shipment(shipment_id: str, logistics_provider_id: str,
                     proposed_route: str, estimated_cost: float) -> dict:
    """Logistics provider proposes a shipment with route and cost."""
    try:
        db = get_firestore_client()
        ship_ref = db.collection("shipments").document(shipment_id)
        ship_doc = ship_ref.get()

        if not ship_doc.exists:
            raise ValueError("Shipment not found")

        shipment = ship_doc.to_dict()
        if shipment["status"] != "awaiting_logistics":
            raise ValueError("Shipment is not awaiting logistics")
        if shipment.get("logisticsProviderId") and shipment["logisticsProviderId"] != logistics_provider_id:
            raise ValueError("Another logistics provider already proposed")

        distance_km = round(random.uniform(50, 350), 1)
        hours = round(distance_km / 45)
        estimated_duration = f"{hours}h {random.randint(5, 59)}m"

        ship_ref.update({
            "logisticsProviderId": logistics_provider_id,
            "status": "logistics_proposed",
            "proposedRoute": proposed_route,
            "estimatedCost": estimated_cost,
            "distanceKm": distance_km,
            "estimatedDuration": estimated_duration,
            "updatedAt": firestore.SERVER_TIMESTAMP,
        })

        logger.info(f"Proposed shipment {shipment_id} by {logistics_provider_id}")
        return {"shipmentId": shipment_id, "status": "logistics_proposed"}

    except Exception as e:
        logger.error(f"propose_shipment failed: {e}")
        raise


def approve_shipment(shipment_id: str, approved: bool) -> dict:
    """Buyer approves or rejects a logistics proposal."""
    try:
        db = get_firestore_client()
        ship_ref = db.collection("shipments").document(shipment_id)
        ship_doc = ship_ref.get()

        if not ship_doc.exists:
            raise ValueError("Shipment not found")

        shipment = ship_doc.to_dict()
        if shipment["status"] != "logistics_proposed":
            raise ValueError("Shipment is not awaiting buyer approval")

        if approved:
            ship_ref.update({
                "status": "approved",
                "updatedAt": firestore.SERVER_TIMESTAMP,
            })
            logger.info(f"Buyer approved shipment {shipment_id}")
            return {"shipmentId": shipment_id, "status": "approved"}
        else:
            ship_ref.update({
                "status": "awaiting_logistics",
                "logisticsProviderId": None,
                "proposedRoute": "",
                "estimatedCost": None,
                "distanceKm": None,
                "estimatedDuration": "",
                "updatedAt": firestore.SERVER_TIMESTAMP,
            })
            logger.info(f"Buyer rejected shipment {shipment_id}")
            return {"shipmentId": shipment_id, "status": "awaiting_logistics"}

    except Exception as e:
        logger.error(f"approve_shipment failed: {e}")
        raise


def update_shipment_status(shipment_id: str, new_status: str) -> dict:
    """Update shipment status with controlled transitions."""
    try:
        db = get_firestore_client()
        ship_ref = db.collection("shipments").document(shipment_id)
        ship_doc = ship_ref.get()

        if not ship_doc.exists:
            raise ValueError("Shipment not found")

        shipment = ship_doc.to_dict()
        current_status = shipment["status"]

        allowed = VALID_TRANSITIONS.get(current_status, [])
        if new_status not in allowed:
            raise ValueError(f'Cannot transition from "{current_status}" to "{new_status}"')

        ship_ref.update({
            "status": new_status,
            "updatedAt": firestore.SERVER_TIMESTAMP,
        })

        if new_status == "delivered" and shipment.get("commitmentId"):
            try:
                commit_ref = db.collection("commitments").document(shipment["commitmentId"])
                commit_doc = commit_ref.get()
                if commit_doc.exists and commit_doc.to_dict().get("status") == "accepted":
                    commit_ref.update({
                        "status": "completed",
                        "updatedAt": firestore.SERVER_TIMESTAMP,
                    })
                    logger.info(f"Commitment {shipment['commitmentId']} → completed")
            except Exception as e:
                logger.error(f"Failed to update commitment status: {e}")

        logger.info(f"Updated shipment {shipment_id} → {new_status}")
        return {"shipmentId": shipment_id, "status": new_status}

    except Exception as e:
        logger.error(f"update_shipment_status failed: {e}")
        raise


def get_available_shipments() -> list:
    try:
        db = get_firestore_client()
        docs = db.collection("shipments").where("status", "==", "awaiting_logistics").get()
        return [{"id": d.id, **d.to_dict()} for d in docs]
    except Exception as e:
        logger.error(f"get_available_shipments failed: {e}")
        return []


def get_shipments_by_provider(provider_id: str) -> list:
    try:
        db = get_firestore_client()
        docs = db.collection("shipments").where("logisticsProviderId", "==", provider_id).get()
        return [{"id": d.id, **d.to_dict()} for d in docs]
    except Exception as e:
        logger.error(f"get_shipments_by_provider failed: {e}")
        return []


def get_shipments_by_farmer(farmer_id: str) -> list:
    try:
        db = get_firestore_client()
        docs = db.collection("shipments").where("farmerId", "==", farmer_id).get()
        return [{"id": d.id, **d.to_dict()} for d in docs]
    except Exception as e:
        logger.error(f"get_shipments_by_farmer failed: {e}")
        return []


def get_shipments_by_buyer(buyer_id: str) -> list:
    try:
        db = get_firestore_client()
        docs = db.collection("shipments").where("buyerId", "==", buyer_id).get()
        return [{"id": d.id, **d.to_dict()} for d in docs]
    except Exception as e:
        logger.error(f"get_shipments_by_buyer failed: {e}")
        return []


def generate_route(shipment_id: str) -> dict:
    """Generate route data for a shipment (mock coordinates + polyline)."""
    try:
        db = get_firestore_client()
        ship_ref = db.collection("shipments").document(shipment_id)
        ship_doc = ship_ref.get()

        if not ship_doc.exists:
            raise ValueError("Shipment not found")

        shipment = ship_doc.to_dict()

        # Prevent double generation when route already exists
        if shipment.get("distanceKm") and shipment.get("routePolyline"):
            logger.info(f"Route already generated for shipment {shipment_id}")
            return {"shipmentId": shipment_id, **shipment, "id": shipment_id}

        # Mock coordinate generation (India bounding box roughly)
        pickup_lat = round(random.uniform(15.0, 25.0), 6)
        pickup_lng = round(random.uniform(73.0, 85.0), 6)
        delivery_lat = round(random.uniform(15.0, 25.0), 6)
        delivery_lng = round(random.uniform(73.0, 85.0), 6)

        distance_km = round(random.uniform(50, 450), 1)
        hours = round(distance_km / 45)
        mins = random.randint(5, 55)
        estimated_duration = f"{hours}h {mins}m"

        # Generate simple mock polyline (list of {lat, lng} objects — Firestore doesn't allow nested arrays)
        num_waypoints = random.randint(4, 8)
        route_polyline = []
        for i in range(num_waypoints):
            frac = i / (num_waypoints - 1)
            lat = round(pickup_lat + frac * (delivery_lat - pickup_lat) + random.uniform(-0.5, 0.5), 6)
            lng = round(pickup_lng + frac * (delivery_lng - pickup_lng) + random.uniform(-0.5, 0.5), 6)
            route_polyline.append({"lat": lat, "lng": lng})

        update_data = {
            "pickupCoordinates": {"lat": pickup_lat, "lng": pickup_lng},
            "deliveryCoordinates": {"lat": delivery_lat, "lng": delivery_lng},
            "routePolyline": route_polyline,
            "distanceKm": distance_km,
            "estimatedDuration": estimated_duration,
            "updatedAt": firestore.SERVER_TIMESTAMP,
        }

        ship_ref.update(update_data)
        logger.info(f"Generated route for shipment {shipment_id}: {distance_km} km, {estimated_duration}")

        result = {**shipment, **update_data, "id": shipment_id}
        return result

    except Exception as e:
        logger.error(f"generate_route failed: {e}")
        raise

